### PR TITLE
First implementation of Hive Metastore

### PR DESCRIPTION
**Pull request by maltesander**
This is a first implementation for Hive.

fixes https://github.com/stackabletech/issues/issues/110

Working:
- [x] Deployment of metastore
- [x] Monitoring for metastore
- [x] Product-config integration
- [x] Restart, Start and Stop Commands
- [x] Versioning working (we only provided one version for now)
- [x] Sticky scheduling
- [x] Tested with derby and postgresql

### CODE REVIEW

File: rust/operator/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,672 @@
...
+    ///
+    async fn create_config_maps(
+        &self,
+        pod_id: &PodIdentity,
+        validated_config: &HashMap<PropertyNameKind, BTreeMap<String, String>>,
+    ) -> Result<HashMap<&'static str, ConfigMap>, Error> {
+        let mut config_maps = HashMap::new();
+        let mut config_maps_data = BTreeMap::new();
+
+        let log4j_config = include_str!("log4j.properties");
```
Comment (soenkeliebau) [inline rust/operator/src/lib.rs:263]: This was just something that I initially implemented to get this of the ground quickly, not sure if we should merge this tbh.
How are we treating log4j files for the other products?

File: rust/operator-binary/Cargo.lock [hunk 2]
```diff
@@ -0,0 +1,1991 @@
+# This file is automatically @generated by Cargo.
```
Comment (soenkeliebau) [inline rust/operator-binary/Cargo.lock:1]: I don't think we need .lock files in the individual crates, the top level one for the workspace Cargo.toml should contain everything, no?

File: rust/operator-binary/src/stackable-hive-operator.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    // Handle CLI arguments
+    let matches = App::new(built_info::PKG_DESCRIPTION)
+        .author("Stackable GmbH - info@stackable.de")
+        .about(built_info::PKG_DESCRIPTION)
+        .version(crate_version!())
+        .arg(cli::generate_productconfig_arg())
+        .subcommand(
+            SubCommand::with_name("crd")
+                .setting(AppSettings::ArgRequiredElseHelp)
+                .subcommand(cli::generate_crd_subcommand::<HiveCluster>()),
```
Comment (soenkeliebau) [inline rust/operator-binary/src/stackable-hive-operator.rs:27]: This is missing the subcommands for start,stop, restart.

File: rust/operator-binary/src/stackable-hive-operator.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        built_info::BUILT_TIME_UTC,
+        built_info::RUSTC_VERSION,
+    );
+
+    let client = client::create_client(Some("hive.stackable.tech".to_string())).await?;
+
+    if let Err(error) = stackable_operator::crd::wait_until_crds_present(
+        &client,
+        vec![
+            HiveCluster::crd_name(),
```
Comment (soenkeliebau) [inline rust/operator-binary/src/stackable-hive-operator.rs:66]: This should also check for the databaseconnection object.

File: rust/crd/src/discovery.rs [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+            pod.metadata
+                .labels
+                .as_ref()
+                .and_then(|labels| labels.get(APP_COMPONENT_LABEL))
+                == Some(name_node_str)
+        })
+        .collect();
+
+    if filtered_pods.len() > 1 {
+        warn!("Retrieved more than one name_node pod. This is not supported and may lead to untested side effects. \
```
Comment (soenkeliebau) [inline rust/crd/src/discovery.rs:144]: Namenode?

File: rust/crd/src/discovery.rs [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+                port,
+                db: cleaned_db_root,
+            }));
+        }
+    }
+
+    Ok(None)
+}
+
+/// Build a Labelselector that applies only to name node pods belonging to the cluster instance
```
Comment (soenkeliebau) [inline rust/crd/src/discovery.rs:179]: Comment still refers to namenode

File: rust/operator/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,672 @@
...
+        Ok(ReconcileFunctionAction::Continue)
+    }
+
+    pub async fn create_missing_pods(&mut self) -> HiveReconcileResult {
+        trace!(target: "create_missing_pods","Starting `create_missing_pods`");
+
+        // The iteration happens in two stages here, to accommodate the way our operators think
+        // about roles and role groups.
+        // The hierarchy is:
+        // - Roles (Master, Worker, History-Server)
```
Comment (soenkeliebau) [inline rust/operator/src/lib.rs:128]: Wrong roles

File: rust/operator/src/lib.rs (hunk 1 above)
Comment (maltesander) [reply rust/operator/src/lib.rs:263]: This only came up with the hdfs and hive. What would you suggest?

**Pull request merged**